runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,455 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+         // Now store the value onto the stack.
+         generateRXInstruction(cg, mnemonic, node, valueReferenceReg, valueMR);
+         }
+      else if (valueReferenceReg->getKind() == TR_FPR)
+         {
+         auto mnemonic = nodeType == TR::Float ? TR::InstOpCode::STE : TR::InstOpCode::STD;
+         // Now store the value onto the stack.
+         generateRXInstruction(cg, mnemonic, node, valueReferenceReg, valueMR);
+         }
+      else
+         {
+         TR_ASSERT_FATAL(false, "Unsupported register kind (%d) for fieldwatch.", valueReferenceReg->getKind());
+         }
+      valueReferenceReg = cg->allocateRegister();
+
+      // Now load the memory location back into the register so that it can be used
+      // as an argument register for the VM helper call.
+      TR::MemoryReference *tempMR = generateS390MemoryReference(*valueMR, 0, cg);
+      generateRXInstruction(cg, TR::InstOpCode::LA, node, valueReferenceReg, tempMR);
+
+      dependencies->addPreCondition(valueReferenceReg, helperLink->getIntegerArgumentRegister(numArgs));
+      dependencies->addPostCondition(valueReferenceReg, helperLink->getIntegerArgumentRegister(numArgs));
+      valueReferenceReg->setPlaceholderReg();
+
+      cg->stopUsingRegister(valueReferenceReg);
+      }
+
+   // These registers will hold Entry Point and Return Address registers, which are required when generating a directCall.
+   TR::Register *scratch1 = cg->allocateRegister();
+   TR::Register *scratch2 = cg->allocateRegister();
+   dependencies->addPostCondition(scratch1, cg->getEntryPointRegister());
+   dependencies->addPostCondition(scratch2, cg->getReturnAddressRegister());
+
+   // Now generate the call to VM Helper to report the fieldwatch.
+   TR::Instruction *call = generateDirectCall(cg, node, false /*myself*/, cg->symRefTab()->findOrCreateRuntimeHelper(helperIndex, false, false, false), dependencies);
+   call->setNeedsGCMap(0x0000FFFF);
+   call->setDependencyConditions(dependencies);
+
+   // After returning from the VM Helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergePointLabel);
+   // End of OOL.
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+
+   cg->stopUsingRegister(scratch1);
+   cg->stopUsingRegister(scratch2);
+
+   cg->stopUsingRegister(dataBlockReg);
+   }
+
+TR::Snippet *
+J9::Z::TreeEvaluator::createDataSnippetForFieldWatch(TR::CodeGenerator *cg, void *dataBlock, TR::Node *node, bool isStatic)
+   {
+   TR::S390WritableDataSnippet *dataSnippet;
+   if (isStatic)
+      {
+      dataSnippet = static_cast<TR::S390WritableDataSnippet *>(cg->CreateConstant(node, dataBlock, sizeof(J9JITWatchedStaticFieldData), true));
+      }
+   else
+      {
+      dataSnippet = static_cast<TR::S390WritableDataSnippet *>(cg->CreateConstant(node, dataBlock, 1 << (int) (log2(sizeof(J9JITWatchedInstanceFieldData)-1)+1), true));
```

Comment:
This looks to be attempting to create an aligned data snippet? Why is this necessary? Snippets are automatically aligned to the size of an address.

Reply:
Don't snippet sizes need to be a power of 2 on Z? IIRC when we last talked about this, it was still a requirement. 

When @r30shah and I were looking at this earlier, the body of the snippet wasn't being emitted because the size wasn't a power of 2. We do something similar in the `genInstanceOfDynamicCacheAndHelperCall` :
https://github.com/eclipse/openj9/blob/6bf7e0de73f5e1ce5daac1d709d07fca97733900/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L5780-L5793

Reply:
See [1]. We already ensure snippets are generated at an 8 byte boundary to ensure stores to snippet locations are atomic. I don't think the above code is necessary either. @r30shah FYI do we need cleanup above as well?

[1] https://github.com/eclipse/omr/blob/9cea1db767a42e54f9381edf665e75808af8fcae/compiler/z/codegen/OMRCodeGenerator.cpp#L4048-L4102

Reply:
@fjeremic  I have opened up issue regarding to this https://github.com/eclipse/omr/issues/1815.
Right now we just bluntly discard snippets which are not power of 2, mostly to keep alignment but this is not clear and that is why I opened up issue to know reasons. 
Basically when someone tried allocate a data snippet which is not power of 2 we need either a `TR_ASSERT_FATAL` or add padding to make it power of 2.

Reply:
It is this guy which discards snippets with size (not power of 2) https://github.com/eclipse/omr/blob/9cea1db767a42e54f9381edf665e75808af8fcae/compiler/z/codegen/OMRCodeGenerator.cpp#L4066 which defined as following. 

```
   // Assume constants will be emitted in order of decreasing size.  Constants should be aligned
   // according to their size.
   //
   // Constant snippet size (S) is a power of 2
   //     - handle when: S = pos(2,exp)
   //     - snippets alignment by size, no max
   #define HANDLE_CONSTANT_SNIPPET(cursor, pow2Size) \
                      (cursor->getConstantSize() == pow2Size
```

Reply:
I see, thanks. I wasn't tracking the above issue. I've tagged it appropriately and referenced the issue here. We'll fix this in the near future. Thanks for helping me understand everyone.